---
title: "WGCNA"
author: "Pieter Clauw"
date: "03/05/2019"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
library(tximport)
library(DESeq2)
library(vsn)
library(limma)
library(ggplot2)
source('~/Desktop/CoExpression New/Scripts/Rfunctions.r')
knitr::opts_knit$set(root.dir = "/Volumes/nordborg/pub/forPieter/WGCNA/")
options(stringsAsFactors = F)
```

## WGCNA 6C and 16C
Script to define co-expression of genes in multiple accessions and replicates in 6C and 16C


```{r load data, cache = T}
samples<- read.table('/Volumes/nordborg/pub/forPieter/WGCNA/samples.txt', header = T, comment.char = '', sep = '\t')
Araport11 <- read.table('/Volumes/nordborg/pub/forPieter/WGCNA/Araport11_GFF3_genes_transposons.201606.ChrM_ChrC_FullName.gtf')
```

```{r read in count data}
# select samples
samples <- samples[!samples$temperature %in% c('6C?', '16C?'), ]
samples <- samples[samples$genoCheck == 'ok', ]
# prepare list of samples
samples$accession <- as.factor(samples$accession)
samples$temperature <- as.factor(samples$temperature)
samples$replicate <- as.factor(samples$replicate)
# get the filenames of selected samples
files <- file.path('/Volumes/nordborg/pub/forPieter/WGCNA/SalmonQuantification/QuasiMap_pseudoG_Trimmed_illumina/', samples$basename, '' ,'_quasiMap_pseudoG_Trimmed_illumina/quant.sf', fsep = '')
names(files) <- samples$sample

# prepare Araport11 data
colnames(Araport11) <- c('chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes')
Araport11$attributes <- as.character(Araport11$attributes)

# create transcript to gene annotation dataframe
tx2gene <- data.frame('transcriptID' = Araport11[,10], 'geneID' = Araport11[,13])

# import count data for selected samples
txi <- tximport(files, type = 'salmon', tx2gene = tx2gene, dropInfReps = T)
dds.full <- DESeqDataSetFromTximport(txi, colData = samples, design = ~  replicate + accession + temperature + accession:temperature)
dds <- estimateSizeFactors(dds.full)
# select only genes with more than in 10 reads in minimum 10% of the samples
idx <- rowSums(counts(dds) >= 10 ) >= nrow(samples) * 0.1
dds<- dds[idx,]
```


## Variance stabilisation
RNA-seq coutn data is typically heteroscedastic, meansing that the variance correlates with the mean.
The variance stabilisation will make the count data more homoscedastic.
This cannot be done for differential expression analyis,
but is advised for clustering, PCA and visualisation.
For more details see [DESeQ2 preprint](https://www.biorxiv.org/content/biorxiv/early/2014/02/19/002832.full.pdf).

```{r compare VST to rlog}
dds.varStab <- varianceStabilizingTransformation(dds, blind=F)
dds.rlog <- rlog(dds, blind=F)


par(mfrow = c(1,2))
meanSdPlot(assay(dds.varStab))
meanSdPlot(assay(dds.rlog))
```
From the plots showing the standard deviation versus the mean of each gene, it is clear that the variance stabilizing transformation performs slighly better in reducing the heteroscedascity of the data compared to the tlog transformation.
Therefore we continue with the variance stabilizing transformed data.

## remove batch effect
```{r check for batch effects}
# PCA plot
pcaData <- plotPCA(dds.varStab, intgroup=c("replicate", "temperature"), returnData=T)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=replicate, shape=temperature)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```
The replicate effect is not obvious in the data.
PC1 is mainly splitting up in the two temperatures
PC2 is a split up between Col-0 (6909) and the rest.

```{r remove batch effect}
dds.varStab.NB <- dds.varStab
assay(dds.varStab.NB) <- removeBatchEffect(assay(dds.varStab.NB), dds.varStab.NB$replicate)

pcaData <- plotPCA(dds.varStab.NB, intgroup=c("temperature", "accession"), returnData=T)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=temperature, shape=accession)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```
After removing batch effect, the temperature split within each accession is more clearly defined.
Therefore we continue with the batch correct data.


## WGCNA
Step-by-step network construction
```{r data for WGCNA}
counts.varStab.NB <- t(assay(dds.varStab.NB))
expr.6C <- counts.varStab.NB[as.character(samples$sample[samples$temperature == '6C']), ]
expr.16C <- counts.varStab.NB[as.character(samples$sample[samples$temperature == '16C']), ]
# combine data of both temperatures in a list
nSets <- 2
expr<- vector(mode = "list", length = nSets)
expr[[1]] <- list(data = as.data.frame(expr.6C))
expr[[2]] <- list(data = as.data.frame(expr.16C))
names(expr) <- c('6C', '16C')
```

```{r library change, include = F}
detach("package:DESeq2", unload = T)
detach("package:tximport", unload = T)
library(WGCNA)
```

```{r data QC}
#checkSets(expr)
#par(mfrow=c(6,4))
# QQ scatter plots, chekc normal distribution of expressiond aat within each samples
for (i in 1:nSets)
{
  readmatrix <- expr[[i]]$data
  for (j in 1:nrow(readmatrix))
  {
    sampleiwant <- t(readmatrix[j,])
    qqnorm(sampleiwant, pch = 1, frame = FALSE, main = paste(names(expr[i]), 'sample', row.names(readmatrix)[j], sep = ' '))
    qqline(sampleiwant, col = "red", lwd = 2)
  }
}

```
For each sample, the expression seems normally distributed, based on QQ-plot.

```{r filter genes}
# filter out genes with too many missing values and zero variance across samples.
# for consensus, both datasets need to be checked simultaneously. Use geneFilterMS.
expr.filter <- vector(mode = "list", length = nSets)
for (i in 1:nSets)
{
  expr.filter[[i]] <- list(data = geneFilter(expr[[i]]$data))
}
names(expr.filter) <- names(expr)
checkSets(expr.filter, checkStructure = T)
```

```{r pick soft threshold}
for (i in 1:nSets)
{
  
  plotSoftThresholdChoices(expr.filter[[i]]$data, maxSoftThrs = 30, title = names(expr.filter)[i])
}
```

```{r set softThreshold}
expr.filter[[1]]$softThreshold <- 6
expr.filter[[2]]$softThreshold <- 20
```
Based on scale-free topology we have chosen two different soft thresholds.
For the 6C expression data, a soft threshold of `r expr.filter[[1]]$softThreshold` seems to be the better choice.
Whie for teh 16C data set, we opt for a soft threshold of `r expr.filter[[2]]$softThreshold`.


```{r adjacencies}
for (i in 1:nSets)
{
  expr.filter[[i]]$adjacency <- abs(bicor(expr.filter[[i]]$data, use = "p", maxPOutliers = 0.10))^expr.filter[[i]]$softThreshold 
}
```
## calculating topological overlap matrices
Warning: This step is memory heavy.
Sometimes R gives an rrror: vector memory exhausted (limit reached?)
Memoty allocated to R in Rstudio can be increased.
Instructions:
in console:
touch .Renviron
open .Renviron

add to this file:
R_MAX_VSIZE=20Gb

20Gb was succesful on a 16Gb mac.
Restart Rstudio

```{r TOM}
for (i in 1:nSets)
{
  expr.filter[[i]]$TOM <- TOMsimilarity(expr.filter[[i]]$adjacency)
  # prevent too much memory usage by not saving the dissimilarity matrix but create it when needed with 1-TOM
#  expr.filter[[i]]$dissTOM <- 1 - expr.filter[[i]]$TOM
}
```

```{r dendrograms}
collectGarbage()
for (i in 1:nSets)
{
  # hierarchical clustering
  expr.filter[[i]]$geneTree <- hclust(as.dist(1- expr.filter[[i]]$TOM), method = "average")
  # plot dendrogram
  sizeGrWindow(12,9)
  plot(expr.filter[[i]]$geneTree, xlab="", sub="", main = paste("Gene clustering on TOM-based dissimilarity for", names(expr.filter[i])), labels = FALSE, hang = 0.04)
}
```


```{r get modules}
for (i in 1:nSets)
{
  # We like large modules, so we set the minimum module size relatively high:
  minModuleSize <- 10
  # Module identification using dynamic tree cut:
  expr.filter[[i]]$dynamicMods <- cutreeDynamic(dendro = expr.filter[[i]]$geneTree, distM = 1 - expr.filter[[i]]$TOM,
  deepSplit = 1, pamRespectsDendro = FALSE,
  minClusterSize = minModuleSize)
  table(expr.filter[[i]]$dynamicMods)
}
```

```{r size distribution unmerged}
for (i in 1:nSets)
{
  hist(expr.filter[[i]]$dynamicMods, main = names(expr.filter[i]))
}
```





```{r plot modules}
for (i in 1:nSets)
{
  # Convert numeric lables into colors
  expr.filter[[i]]$dynamicColors <- labels2colors(expr.filter[[i]]$dynamicMods)
  # table(dynamicColors)
  # Plot the dendrogram and colors underneath
  sizeGrWindow(8,6)
  plotDendroAndColors(expr.filter[[i]]$geneTree, expr.filter[[i]]$dynamicColors, "Dynamic Tree Cut",
  dendroLabels = FALSE, hang = 0.03,
  addGuide = TRUE, guideHang = 0.05,
  main = paste("Gene dendrogram and module colors for", names(expr.filter)[i]))
}
```

# Merge clusters that are similar

Before, we plot the clustering of module eigengenes to decide the level where to merge

```{r decide level to merge}
for (i in 1:nSets)
{
  # Calculate eigengenes
  MEList <- moduleEigengenes(expr.filter[[i]]$data, colors=  expr.filter[[i]]$dynamicColors)
  MEs <- MEList$eigengenes
  # Calculate dissimilarity of module eigengenes
  MEDiss = 1-cor(expr.filter[[i]]$MEs)
  # Cluster module eigengenes
  METree = hclust(as.dist(MEDiss), method = "average")
  # Plot the result
  sizeGrWindow(7, 6)
  plot(METree, main = paste("Clustering of module eigengenes", names(expr.filter)[i]),
  xlab = "", sub = "", cex=0.3)
  
  MEDissThres <- 0.15
  # Plot the cut line into the dendrogram
  abline(h=MEDissThres, col = "red")
}
```

Now we can merge 

```{r merge similar clusters}
#this step is different comparing to the tutorial for set-specific network construction and it's taken from the tutorial for consensus as it seems to work better (with the first one the moduleLabels it's not giving the same number of modules as the moduleColors, don't know why)

for (i in 1:nSets)
{
  # Call an automatic merging function
  merge <- mergeCloseModules(expr.filter[[i]]$data, expr.filter[[i]]$dynamicMods, cutHeight = MEDissThres, verbose = 3)
  # The merged module labels
  expr.filter[[i]]$mergedLabels <- merge$colors
  # translate into module colors
  expr.filter[[i]]$mergedColors <- labels2colors(expr.filter[[i]]$mergedLabels)
  # Eigengenes of the new merged modules
  expr.filter[[i]]$MEs <- merge$newMEs
  
  sizeGrWindow(12, 9)
  plotDendroAndColors(expr.filter[[i]]$geneTree, cbind(expr.filter[[i]]$dynamicColors, expr.filter[[i]]$mergedColors),
  c("Dynamic Tree Cut", "Merged dynamic"),
  dendroLabels = FALSE, hang = 0.03,
  addGuide = TRUE, guideHang = 0.05,
  main = paste(names(expr.filter)[i]))
}

```


```{r size distribution merged}
for (i in 1:nSets)
{
  hist(expr.filter[[i]]$mergedLabels, main = names(expr.filter[i]))
}
```

## Check quality of modules based on connectivity
```{r connectivity}

# TeMP function
#start with gene connectivity
for (i in 1:nSets) {
  expr.filter[[i]]$geneConnectivity <- intramodularConnectivity(expr.filter[[i]]$adjacency, expr.filter[[i]]$mergedLabels, scaleByMax = FALSE)
  # summarize per module
  modules <- unique(na.omit(expr.filter[[i]]$mergedColors))
  table <- data.frame('Module name'= character(), 'KWithin'= integer(), 'KOut'= integer(), 'Quality module'=integer())
  expr.filter[[i]]$moduleConnectivity <- table
  expr.filter[[i]]$moduleConnectivity<- summarizeConnectivity(expr.filter[[i]], modules=modules)
}

```

Module quality

```{r module quality distribution}
for (i in 1:nSets)
{
  hist(expr.filter[[i]]$moduleConnectivity$Quality.module, main = names(expr.filter[i]))
}

```

## Find overlaps between 6C and 16C

```{r overlap 6C and 16C}
for (i in 1:nSets) {
  expr.filter[[i]]$MEs<- orderMEs(expr.filter[[i]]$MEs, greyName = 'MEO')
  # Isolate the module labels in the order they appear in ordered module eigengenes
  ModuleLabels<- substring(names(expr.filter[[i]]$MEs), 3)
  # Convert the numeric module labels to color labels
  expr.filter[[i]]$Modules<- labels2colors(as.numeric(ModuleLabels))
}
# Initialize tables of p-values and of the corresponding counts
pTable<- matrix(0, nrow = length(expr.filter[[1]]$Modules), ncol = length(expr.filter[[2]]$Modules))
CountTbl<- matrix(0, nrow = length(expr.filter[[1]]$Modules), ncol = length(expr.filter[[2]]$Modules))

# Execute all pairwaise comparisons
for (mod6 in 1:length(expr.filter[[1]]$Modules)) {
  for (mod16 in 1:length(expr.filter[[2]]$Modules)) {
    Members6 <- (expr.filter[[1]]$mergedColors == expr.filter[[1]]$Modules[mod6])
    Members16 <- (expr.filter[[2]]$mergedColors == expr.filter[[2]]$Modules[mod16])
     pTable[mod6, mod16] = -log10(fisher.test(Members6, Members16, alternative = "greater")$p.value);
    CountTbl[mod6, mod16] = sum(expr.filter[[1]]$mergedColors == expr.filter[[1]]$Modules[mod6] & expr.filter[[2]]$mergedColors == expr.filter[[2]]$Modules[mod16])
  }
}

#create color-coded table of the intersection counts
Heatmap_overlap(pTable = pTable, CountTbl = CountTbl, expr = expr.filter)

#the figure it's too big, i will add soon the script to create a table which shows everything clearer 
```

Create a table which describes better the overlaps, use Bonferroni correction

```{r clear table describing overlaps between 6C and 16C}

#set a threshold to visualize the overlaps with Bonferroni correction
Threshold<- -log10(0.05/((length(expr.filter[[1]]$Modules))*(length(expr.filter[[2]]$Modules))))

ClearTable<- data.frame('NameModule16'= character(), 'NameModule6'=character(), 'N_totgenes16'=integer(), 'N_totgenes6'=integer(), 'N_overlapped_genes'=integer(), 'p-value'=integer())

ClearTable<-Table.overlap(expr = expr.filter, pTable = pTable, Threshold = Threshold, CountTbl = CountTbl)

kable(ClearTable)
```





