---
title: "WGCNA"
author: "Pieter Clauw"
date: "03/05/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tximport)
library(DESeq2)
library(vsn)
library(limma)
library(ggplot2)
source('~/tmp/coExpression/Scripts/Rfunctions.r')
knitr::opts_knit$set(root.dir = "/Volumes/nordborg/pub/forPieter/WGCNA/")
options(stringsAsFactors = F)
```

## WGCNA 6C and 16C
Script to define co-expression of genes in multiple accessions and replicates in 6C and 16C


```{r load data, cache = T}
samples<- read.table('/Volumes/nordborg/pub/forPieter/WGCNA/samples.txt', header = T, comment.char = '', sep = '\t')
Araport11 <- read.table('/Volumes/nordborg/pub/forPieter/WGCNA/Araport11_GFF3_genes_transposons.201606.ChrM_ChrC_FullName.gtf')
```

```{r read in count data}
# select samples
samples <- samples[!samples$temperature %in% c('6C?', '16C?'), ]
samples <- samples[samples$genoCheck == 'ok', ]
# prepare list of samples
samples$accession <- as.factor(samples$accession)
samples$temperature <- as.factor(samples$temperature)
samples$replicate <- as.factor(samples$replicate)
# get teh filenames of selected samples
files <- file.path('/Volumes/nordborg/pub/forPieter/WGCNA/SalmonQuantification/QuasiMap_pseudoG_Trimmed_illumina/', samples$basename, '' ,'_quasiMap_pseudoG_Trimmed_illumina/quant.sf', fsep = '')
names(files) <- samples$sample

# prepare Araport11 data
colnames(Araport11) <- c('chr', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes')
Araport11$attributes <- as.character(Araport11$attributes)

# create transcript to gene annotation dataframe
tx2gene <- data.frame('transcriptID' = Araport11[,10], 'geneID' = Araport11[,13])

# import count data for selected samples
txi <- tximport(files, type = 'salmon', tx2gene = tx2gene, dropInfReps = T)
dds.full <- DESeqDataSetFromTximport(txi, colData = samples, design = ~  replicate + accession + temperature + accession:temperature)
dds <- estimateSizeFactors(dds.full)
# select only genes with more than in 10 reads in minimum 10% of the samples
idx <- rowSums(counts(dds) >= 10 ) >= nrow(samples) * 0.1
dds<- dds[idx,]
```


## Variance stabilisation
RNA-seq coutn data is typically heteroscedastic, meansing that the variance correlates with the mean.
The variance stabilisation will make the count data more homoscedastic.
This cannot be done for differential expression analyis,
but is advised for clustering, PCA and visualisation.
For more details see [DESeQ2 preprint](https://www.biorxiv.org/content/biorxiv/early/2014/02/19/002832.full.pdf).

```{r compare VST to rlog}
dds.varStab <- varianceStabilizingTransformation(dds, blind=F)
dds.rlog <- rlog(dds, blind=F)


par(mfrow = c(1,2))
meanSdPlot(assay(dds.varStab))
meanSdPlot(assay(dds.rlog))
```
From the plots showing the standard deviation versus the mean of each gene, it is clear that the variance stabilizing transformation performs slighly better in reducing the heteroscedascity of the data compared to the tlog transformation.
Therefore we continue with the variance stabilizing transformed data.

## remove batch effect
```{r check for batch effects}
# PCA plot
pcaData <- plotPCA(dds.varStab, intgroup=c("replicate", "temperature"), returnData=T)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=replicate, shape=temperature)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```
The replicate effect is not obvious in the data.
PC1 is mainly splitting up in the two temperatures
PC2 is a split up between Col-0 (6909) and the rest.

```{r remove batch effect}
dds.varStab.NB <- dds.varStab
assay(dds.varStab.NB) <- removeBatchEffect(assay(dds.varStab.NB), dds.varStab.NB$replicate)

pcaData <- plotPCA(dds.varStab.NB, intgroup=c("temperature", "accession"), returnData=T)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=temperature, shape=accession)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
counts.varStab.NB <- t(assay(dds.varStab.NB))
```
After removing batch effect, the temperature split within each accession is more clearly defined.
Therefore we continue with the batch correct data.


## WGCNA
Step-by-step network construction
```{r data for WGCNA}
counts.varStab.NB <- t(assay(dds.varStab.NB))
expr.6C <- counts.varStab.NB[as.character(samples$sample[samples$temperature == '6C']), ]
expr.16C <- counts.varStab.NB[as.character(samples$sample[samples$temperature == '16C']), ]
# combine data of both temperatures in a list
nSets <- 2
expr<- vector(mode = "list", length = nSets)
expr[[1]] = list(data = as.data.frame(expr.6C))
expr[[2]] = list(data = as.data.frame(expr.16C))
```

```{r library change, include = F}
detach("package:DESeq2", unload = T)
detach("package:tximport", unload = T)
library(WGCNA)
```

```{r data QC}
#checkSets(expr)
#par(mfrow=c(6,4))
# QQ scatter plots, chekc normal distribution of expressiond aat within each samples
for (i in 1:nSets)
{
  readmatrix <- expr[[i]]$data
  for (j in 1:nrow(readmatrix))
  {
    sampleiwant <- t(readmatrix[j,])
    qqnorm(sampleiwant, pch = 1, frame = FALSE, main = paste(names(expr[i]), 'sample', row.names(readmatrix)[j], sep = ' '))
    qqline(sampleiwant, col = "red", lwd = 2)
  }
}

```
For each sample, the expression seems normally distributed, based on QQ-plots










